src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -180,6 +208,156 @@ public void Pop(bool success)
+                }
+
+                throw new AggregateException(exceptions);
+            }
+
+            PendingAsyncDisposables.Clear();
+        }
+
+        /// <summary>
+        /// Walks the stack cleaning up any leftover IDisposables
+        /// in the event of an exception on serialization
+        /// </summary>
+        public void DisposePendingDisposablesOnException()
+        {
+            List<Exception>? exceptions = null;
+
+            Debug.Assert(Current.AsyncEnumerator is null);
+            DisposeFrame(Current.CollectionEnumerator);
+
+            int stackSize = Math.Max(_count, _continuationCount);
+            if (stackSize > 1)
+            {
+                for (int i = 0; i < stackSize - 1; i++)
+                {
+                    Debug.Assert(_previous[i].AsyncEnumerator is null);
+                    DisposeFrame(_previous[i].CollectionEnumerator);
+                }
+            }
+
+            if (exceptions is not null)
+            {
+                if (exceptions.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(exceptions[0]).Throw();
+                }
+
+                throw new AggregateException(exceptions);
+            }
+
+            void DisposeFrame(IEnumerator? collectionEnumerator)
+            {
+                try
+                {
+                    if (collectionEnumerator is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                        return;
+                    }
+                }
+                catch (Exception e)
+                {
+                    exceptions ??= new();
+                    exceptions.Add(e);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Walks the stack cleaning up any leftover I(Async)Disposables
+        /// in the event of an exception on async serialization
+        /// </summary>
+        public async ValueTask DisposePendingDisposablesOnExceptionAsync()
+        {
+            List<Exception>? exceptions = null;
+
+            await DisposeFrame(Current.CollectionEnumerator, Current.AsyncEnumerator).ConfigureAwait(false);
+
+            int stackSize = Math.Max(_count, _continuationCount);
+            if (stackSize > 1)
+            {
+                for (int i = 0; i < stackSize - 1; i++)
+                {
+                    await DisposeFrame(_previous[i].CollectionEnumerator, _previous[i].AsyncEnumerator).ConfigureAwait(false);
+                }
+            }
+
+            if (exceptions is not null)
+            {
+                if (exceptions.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(exceptions[0]).Throw();
+                }
+
+                throw new AggregateException(exceptions);
+            }
+
+            async ValueTask DisposeFrame(IEnumerator? collectionEnumerator, IAsyncDisposable? asyncDisposable)
```

Comment:
Using a local async function that closes over state is likely to lead to more allocations than you might expect.  Take a look at the IL and see what's there, but I expect there's an additional display class allocation in order to handle `exceptions`.

Reply:
Since async methods can't accept ref parameters, if we want to avoid allocating closures we could either a) preallocate `List<Exception>` or b) change the signature into something like `static async ValueTask<List<Exception>?> DisposeFrame(..., List<Exception>? exceptions)` and fold over the disposables.